File: src/main/java/com/hackclub/hccore/DataManager.java
package com.hackclub.hccore;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.function.Predicate;
import java.util.logging.Level;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;

public class DataManager {

  private final HCCorePlugin plugin;
  private final String dataFolder;
  private final Map<UUID, PlayerData> onlinePlayers = new HashMap<>();

  public DataManager(HCCorePlugin plugin) {
    this.plugin = plugin;
    this.dataFolder = plugin.getDataFolder() + File.separator + "players";

    File folder = new File(this.dataFolder);
    folder.mkdirs();
  }

  public String getDataFolder() {
    return this.dataFolder;
  }

  public PlayerData getData(Player player) {
    return this.onlinePlayers.get(player.getUniqueId());
  }

  public PlayerData getData(OfflinePlayer offlinePlayer) {
    PlayerData data = this.onlinePlayers.get(offlinePlayer.getUniqueId());

    if (data == null) {
      data = new PlayerData(this.plugin, offlinePlayer);
      data.load();
    }

    return data;
  }

  public PlayerData findData(Predicate<? super PlayerData> predicate) {
    PlayerData onlineMatch = this.onlinePlayers.values().stream().filter(predicate).findFirst()
        .orElse(null);
    if (onlineMatch != null) {
      return onlineMatch;
    }

    File[] files = new File(this.dataFolder).listFiles(
        (directory, name) -> name.endsWith(".json"));
    if (files == null) {
      return null;
    }

    Set<UUID> onlineUuids = new HashSet<>(this.onlinePlayers.keySet());
    for (File file : files) {
      String filename = file.getName();
      try {
        UUID uuid = UUID.fromString(filename.substring(0, filename.length() - ".json".length()));
        if (onlineUuids.contains(uuid)) {
          continue;
        }

        PlayerData data = this.getData(this.plugin.getServer().getOfflinePlayer(uuid));
        if (predicate.test(data)) {
          return data;
        }
      } catch (IllegalArgumentException exception) {
        this.plugin.getLogger().log(Level.WARNING,
            "Ignoring player data file with an invalid UUID: " + filename);
      }
    }

    return null;
  }

  public List<PlayerData> findDataMany(Predicate<? super PlayerData> predicate) {
    List<PlayerData> onlineMatches = this.onlinePlayers.values().stream().filter(predicate).toList();
    List<PlayerData> matches = new ArrayList<>(onlineMatches);

    File[] files = new File(this.dataFolder).listFiles(
        (directory, name) -> name.endsWith(".json"));
    if (files == null) {
      return matches;
    }

    Set<UUID> onlineUuids = new HashSet<>(this.onlinePlayers.keySet());
    for (File file : files) {
      String filename = file.getName();
      try {
        UUID uuid = UUID.fromString(filename.substring(0, filename.length() - ".json".length()));
        if (onlineUuids.contains(uuid)) {
          continue;
        }

        PlayerData data = this.getData(this.plugin.getServer().getOfflinePlayer(uuid));
        if (predicate.test(data)) {
          matches.add(data);
        }
      } catch (IllegalArgumentException exception) {
        this.plugin.getLogger().log(Level.WARNING,
            "Ignoring player data file with an invalid UUID: " + filename);
      }
    }

    return matches;
  }

  public void registerPlayer(Player player) {
    this.onlinePlayers.put(player.getUniqueId(), new PlayerData(this.plugin, player));

    // Register player's team
    Scoreboard mainScoreboard = player.getServer().getScoreboardManager().getMainScoreboard();
    player.setScoreboard(mainScoreboard);
    // Unregister existing teams in the player's name
    Team playerTeam = mainScoreboard.getTeam(player.getName());
    if (playerTeam != null) {
      playerTeam.unregister();
    }
    playerTeam = mainScoreboard.registerNewTeam(player.getName());
    playerTeam.addEntry(player.getName());

    // Load in player data for use
    this.getData(player).load();
  }

  public void registerAll() {
    for (Player player : this.plugin.getServer().getOnlinePlayers()) {
      this.registerPlayer(player);
    }
  }

  public void unregisterPlayer(Player player) {
    this.getData(player).save();

    this.getData(player).getTeam().unregister();

    this.onlinePlayers.remove(player.getUniqueId());
  }

  public void unregisterAll() {
    for (Player player : this.plugin.getServer().getOnlinePlayers()) {
      this.unregisterPlayer(player);
    }
  }
}

File: src/main/java/com/hackclub/hccore/DataManager.kt
package com.hackclub.hccoreimport

import org.bukkit.entity.Player
import java.io.File

com.hackclub.hccore.HCCorePlugin
import com.hackclub.hccore.PlayerData
import com.hackclub.hccore.tasks.CheckAdAstraTask
import com.hackclub.hccore.utils.gson.GsonUtil
import com.hackclub.hccore.utils.gson.LocationSerializer
import com.hackclub.hccore.utils.gson.LocationMapDeserializer
import com.hackclub.hccore.utils.gson.LocationMapSerializer
import com.hackclub.hccore.utils.gson.LocationDeserializer
import java.util.LinkedHashMap
import com.hackclub.hccore.events.player.PlayerAFKStatusChangeEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.JDABuilder
import javax.security.auth.login.LoginException
import java.lang.Exception
import com.hackclub.hccore.commands.ColorCommand
import java.util.stream.Collectors
import com.hackclub.hccore.commands.StatsCommand
import java.text.SimpleDateFormat
import com.hackclub.hccore.commands.UpvoteCommand
import com.hackclub.hccore.commands.DownvoteCommand
import com.hackclub.hccore.commands.TableflipCommand
import java.lang.Runnable
import com.hackclub.hccore.listeners.SleepListener
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import com.comphenix.protocol.wrappers.EnumWrappers
import com.comphenix.protocol.wrappers.WrappedGameProfile
import org.bukkit.craftbukkit.v1_17_R1.advancement.CraftAdvancement
import net.minecraft.advancements.AdvancementDisplay
import net.minecraft.network.chat.IChatBaseComponent
import net.minecraft.advancements.AdvancementFrameType
import de.tr7zw.changeme.nbtapi.NBTTileEntity
import java.io.FileReader
import java.io.IOException
import java.io.FileWriter
import java.util.UUID
import java.util.HashMap
import com.hackclub.hccore.DataManager
import com.comphenix.protocol.ProtocolManager
import com.hackclub.hccore.discord.DiscordBot
import com.comphenix.protocol.ProtocolLibrary
import com.hackclub.hccore.commands.AFKCommand
import com.hackclub.hccore.commands.LocCommand
import com.hackclub.hccore.commands.NickCommand
import com.hackclub.hccore.commands.PingCommand
import com.hackclub.hccore.commands.ShrugCommand
import com.hackclub.hccore.commands.SpawnCommand
import com.hackclub.hccore.commands.DiscordCommand
import com.hackclub.hccore.listeners.AdvancementListener
import com.hackclub.hccore.listeners.AFKListener
import com.hackclub.hccore.listeners.BeehiveInteractionListener
import com.hackclub.hccore.listeners.PlayerListener
import com.hackclub.hccore.listeners.NameChangeListener
import com.hackclub.hccore.tasks.AutoAFKTask
import hu.trigary.advancementcreator.AdvancementFactory
import hu.trigary.advancementcreator.shared.ItemObject

class DataManager(plugin: HCCorePlugin) {
    private val plugin: HCCorePlugin
    val dataFolder: String
    private val players: MutableMap<UUID, PlayerData> = HashMap()
    fun getData(player: Player): PlayerData? {
        return players[player.uniqueId]
    }

    fun registerPlayer(player: Player) {
        players[player.uniqueId] = PlayerData(plugin, player)

        // Register player's team
        val mainScoreboard = player.server.scoreboardManager!!.mainScoreboard
        player.scoreboard = mainScoreboard
        // Unregister existing teams in the player's name
        var playerTeam = mainScoreboard.getTeam(player.name)
        playerTeam?.unregister()
        playerTeam = mainScoreboard.registerNewTeam(player.name)
        playerTeam.addEntry(player.name)

        // Load in player data for use
        getData(player)!!.load()
    }

    fun registerAll() {
        for (player in plugin.getServer().getOnlinePlayers()) {
            registerPlayer(player)
        }
    }

    fun unregisterPlayer(player: Player) {
        getData(player)!!.save()
        getData(player)!!.team.unregister()
        players.remove(player.uniqueId)
    }

    fun unregisterAll() {
        for (player in plugin.getServer().getOnlinePlayers()) {
            unregisterPlayer(player)
        }
    }

    init {
        this.plugin = plugin
        dataFolder = plugin.getDataFolder().toString() + File.separator + "players"
        val folder = File(dataFolder)
        folder.mkdirs()
    }
}